### ISSUE TITLE
LateError LateInitializationError: Field 'db' has not been initialized.

### ISSUE THREAD

**Issue by phijma** [simolus3/moor#1250]
The reason I "file" a bug is that this was no problem before. :) If a user starts the app and is unauthenticated, I would like to log the user out and remove their database. But since using the new version of flutter + dart which has null safety and such, I get this error:

```
LateError: LateInitializationError: Field 'db' has not been initialized.
  File "encrypted_moor.dart", in _SqfliteDelegate.close
  File "engines.dart", line 310, in DelegatedDatabase.close
  File "db_base.dart", line 128, in GeneratedDatabase.close
  File "<asynchronous suspension>"
  File "app_db.dart", line 101, in AppDB.removeDatabase
  File "<asynchronous suspension>"
  File "authentication_service.dart", line 75, in AuthenticationService.removeUserData
  File "<asynchronous suspension>"
  File "base_screen.dart", line 42, in BaseScreen.build.<fn>
```

Any clue?

This is no problem of course when the user is authenticated and uses the logout button. As in the meantime the database is always initialized. This is only a problem if the user is authenticated, closes the app, and then restarts it.

**Comment by simolus3** [comment]
Thanks for reporting this! Moor should handle closing a database that has never been opened. We have an integration test for this as well, and that does pass with `encrypted_moor`:

https://github.com/simolus3/moor/blob/ac30bb598946b0fbfb9e278e73cf9451c01e237c/extras/integration_tests/tests/lib/suite/suite.dart#L30-L34

Can you try to reproduce this with a simple example that I could take a look at? E.g., does it happen if you create a database (without using  it) and then call `close()` right away?

**Comment by phijma** [comment]
I can! At the moment I don't have time. But I will in the upcoming days.

**Comment by simolus3** [comment]
Yeah the `encrypted_moor` packages uses David's sqlcipher too.

**Comment by phijma** [comment]
My first observation: this does not happen inside the simulator. So that could explain the tests succeed maybe. 

Simple example is very hard because of this observation.

**Comment by phijma** [comment]
Observation: I can reproduce it in the Android simulator. Though it does lead to this error only **sometimes**.

That might explain that the tests do not fail in a way.

**Comment by davidmartos96** [comment]
@phijma Can you try this?
```dart
final appDatabase = AppDatabase();
await appDatabase.close();
```
I'd say the problem is simply that there is no check for the app being open when closing it.

https://github.com/simolus3/moor/blob/ac30bb598946b0fbfb9e278e73cf9451c01e237c/extras/encryption/lib/encrypted_moor.dart#L77

**Comment by phijma** [comment]
> @phijma Can you try this?
> 
> ```dart
> final appDatabase = AppDatabase();
> await appDatabase.close();
> ```
> 
> I'd say the problem is simply that there is no check for the database being open when closing it.
> 
> https://github.com/simolus3/moor/blob/ac30bb598946b0fbfb9e278e73cf9451c01e237c/extras/encryption/lib/encrypted_moor.dart#L77

Hm, that seems to work without a problem! So that is weird.

Maybe I'm doing something completely wrong in combination with riverpod.

**Comment by simolus3** [comment]
We have that check in another wrapper:

https://github.com/simolus3/moor/blob/ac30bb598946b0fbfb9e278e73cf9451c01e237c/moor/lib/src/runtime/executor/helpers/engines.dart#L307-L316

There could technically be a scenario where `ensureOpen` has been called, but didn't complete. Then, `close()` is called and prematurely closes the underlying delegate.

**Comment by phijma** [comment]
@simolus3 

> We have that check in another wrapper:
> 
> https://github.com/simolus3/moor/blob/ac30bb598946b0fbfb9e278e73cf9451c01e237c/moor/lib/src/runtime/executor/helpers/engines.dart#L307-L316
> 
> There could technically be a scenario where `ensureOpen` has been called, but didn't complete. Then, `close()` is called and prematurely closes the underlying delegate.

Would explain that it happens sometimes. (I've also updated my previous comment with a "wild guess")

**Comment by simolus3** [comment]
Apart from the scenario I mentioned, there's another potential problem if `close()` is called multiple times.

Both problems should be fixed with the referenced commit.

**Comment by phijma** [comment]
Thank you for the quick responses and even a fix @simolus3 

 I will try “the commit” tomorrow.

Let me use this moment to also thank you for flutter moor and @davidmartos96 as well for his package! Works very well.

**Comment by phijma** [comment]
@simolus3 Tested and seems to work! No errors anymore on three different simulators.

Any idea when this might be released? Days, weeks or possibly months?

Thanks again.

**Comment by simolus3** [comment]
A couple of weeks is a realistic timeframe I think. I'm a bit busy right now, and the next moor version contains a whole bunch of cool features that I still need to fledge out.

**Comment by flutterq2** [comment]
All possible working solution added [[Solved] LateInitializationError: Field ‘_userData@78458596’ has not been initialized](https://fluttercorner.com/lateinitializationerror-field-userdata78458596-has-not-been-initialized/)

**Comment by peterhijma** [comment]
> All possible working solution added [[Solved] LateInitializationError: Field ‘_userData@78458596’ has not been initialized](https://fluttercorner.com/lateinitializationerror-field-userdata78458596-has-not-been-initialized/)

How is this helping? I don't it does regarding this issue.

**Comment by North101** [comment]
> > All possible working solution added [[Solved] LateInitializationError: Field ‘_userData@78458596’ has not been initialized](https://fluttercorner.com/lateinitializationerror-field-userdata78458596-has-not-been-initialized/)
> 
> How is this helping? I don't think it does regarding this issue.

based on the name and what they are linking to i think its just a spam account that copies stackoverflow (or similar), puts. some ads and then uses a bot to crawl through github and comment on posts linking to their site

**Comment by peterhijma** [comment]
@simolus3 Sorry to bother you about this issue again. But do you consider it safe to temporarily fork the repo, add the fix of this issue to the master branch and use that?

Until a new version is out of course :)

**Comment by North101** [comment]
> @simolus3 Sorry to bother you about this issue again. But do you consider it safe to temporarily fork the repo, add the fix of this issue to the master branch and use that?
> 
> Until a new version is out of course :)

You can use this to use the latest version (change beta to develop)
https://moor.simonbinder.eu/docs/internals/

**Comment by peterhijma** [comment]
> > @simolus3 Sorry to bother you about this issue again. But do you consider it safe to temporarily fork the repo, add the fix of this issue to the master branch and use that?
> > Until a new version is out of course :)
> 
> You can use this to use the latest version (change beta to develop)
> https://moor.simonbinder.eu/docs/internals/

Yeah, but that is a lot more risky I guess? Instead of a stable version with only one change.

**Comment by simolus3** [comment]
I've cherry-picked the fix onto the `fix-close-without-open` branch which is otherwise identical to `master` If you use git dependencies you only need to override `moor` since you'll be using the unchanged version from pub for the other packages.

**Comment by peterhijma** [comment]
> I've cherry-picked the fix onto the `fix-close-without-open` branch which is otherwise identical to `master`. If you use git dependencies you only need to override `moor` since you'll be using the unchanged version from pub for the other packages.

Awesome! Thank you.

**Comment by peterhijma** [comment]
@simolus3 I see it's included in 4.4.0 🥳 So for me: the branch could be closed again

**Comment by simolus3** [comment]
Thanks for the reminder, I've deleted the branch :)
